File: DuDoong-Api/src/main/java/band/gosrock/api/issuedTicket/dto/response/IssuedTicketAdminTableElement.java
package band.gosrock.api.issuedTicket.dto.response;


import band.gosrock.domain.common.vo.IssuedTicketInfoVo;
import band.gosrock.domain.common.vo.IssuedTicketOptionAnswerVo;
import band.gosrock.domain.common.vo.UserInfoVo;
import band.gosrock.domain.domains.issuedTicket.domain.IssuedTicket;
import band.gosrock.domain.domains.issuedTicket.domain.IssuedTicketOptionAnswer;
import band.gosrock.domain.domains.order.domain.Order;
import band.gosrock.domain.domains.user.domain.User;
import com.fasterxml.jackson.annotation.JsonUnwrapped;
import java.util.List;
import lombok.Builder;
import lombok.Getter;

@Getter
@Builder
public class IssuedTicketAdminTableElement {

    @JsonUnwrapped private final IssuedTicketInfoVo issuedTicketInfo;

    private final UserInfoVo userInfo;

    private final String orderNo;

    private final List<IssuedTicketOptionAnswerVo> issuedTicketOptionAnswers;

    public static IssuedTicketAdminTableElement of(
            IssuedTicket issuedTicket, User user, Order order) {
        return IssuedTicketAdminTableElement.builder()
                .userInfo(user.toUserInfoVo())
                .orderNo(order.getOrderNo())
                .issuedTicketInfo(issuedTicket.toIssuedTicketInfoVo())
                .issuedTicketOptionAnswers(
                        issuedTicket.getIssuedTicketOptionAnswers().stream()
                                .map(IssuedTicketOptionAnswer::toIssuedTicketOptionAnswerVo)
                                .toList())
                .build();
    }
}

File: DuDoong-Api/src/main/kotlin/band/gosrock/api/issuedTicket/dto/response/IssuedTicketAdminTableElement.kt
package band.gosrock.api.issuedTicket.dto.response

import band.gosrock.domain.common.vo.IssuedTicketInfoVo
import band.gosrock.domain.common.vo.IssuedTicketOptionAnswerVo
import band.gosrock.domain.common.vo.UserInfoVo
import band.gosrock.domain.domains.issuedTicket.domain.IssuedTicket
import band.gosrock.domain.domains.issuedTicket.domain.IssuedTicketOptionAnswer
import band.gosrock.domain.domains.order.domain.Order
import band.gosrock.domain.domains.user.domain.User
import com.fasterxml.jackson.annotation.JsonUnwrapped

data class IssuedTicketAdminTableElement(
    @JsonUnwrapped val issuedTicketInfo: IssuedTicketInfoVo,
    val userInfo: UserInfoVo,
    val orderNo: String?,
    val issuedTicketOptionAnswers: List<IssuedTicketOptionAnswerVo>,
) {
    companion object {
        @JvmStatic
        fun of(issuedTicket: IssuedTicket, user: User, order: Order): IssuedTicketAdminTableElement {
            return IssuedTicketAdminTableElement(
                issuedTicketInfo = issuedTicket.toIssuedTicketInfoVo(),
                userInfo = user.toUserInfoVo(),
                orderNo = order.orderNo,
                issuedTicketOptionAnswers = issuedTicket.issuedTicketOptionAnswers
                    .map(IssuedTicketOptionAnswer::toIssuedTicketOptionAnswerVo),
            )
        }
    }
}
